src/grid_interpolator.cc

```diff
@@ -47,21 +59,23 @@ GridInterpolator::~GridInterpolator()
 
 double GridInterpolator::operator()(double x, double y, double z) const
     {
-    const double fx = x/dx_;
-    const double fy = y/dy_;
-    const double fz = z/dz_;
-    const int bin_x = std::floor(fx); 
-    const int bin_y = std::floor(fy); 
-    const int bin_z = std::floor(fz); 
+    const double fx = (x - std::get<0>(lower_bounds_))/dx_;
+    const double fy = (y - std::get<1>(lower_bounds_))/dy_;
+    const double fz = (z - std::get<2>(lower_bounds_))/dz_;
+    
+    const auto bounds = n_.getBounds();
+    const int bin_x = (std::abs(std::floor(fx) - (std::get<0>(bounds) - 1)) == 0) ? std::floor(fx) - 1 : std::floor(fx); 
```

Comment:
You are on the right track, but I don't think this is how I would do this. Instead, I would probably do something like:
```suggestion
   int bin_x = std::floor(fx);
   if(fx == std::get<0>(bounds) && x == std::get<0>(upper_bounds_))
       {
       --bin_x;
       }
```
The second condition could be relaxed to an "is close" type tolerance if we find this is not working.